From 589ab3dbf4af3412daa40524024e58f8b548d263 Mon Sep 17 00:00:00 2001
From: Maximilian Luz <luzmaximilian@gmail.com>
Date: Fri, 6 Mar 2020 17:55:15 +0100
Subject: [PATCH 2/3] ipts: Do not forward report on overflow

If an overflow is encountered caused by a too large HID input, do not
forward it to the HID subsystem. Instead drop it and emit an error
message.
---
 drivers/misc/ipts/hid.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/misc/ipts/hid.c b/drivers/misc/ipts/hid.c
index 3652803b5376..c3e4ef9640e8 100644
--- a/drivers/misc/ipts/hid.c
+++ b/drivers/misc/ipts/hid.c
@@ -303,8 +303,11 @@ int ipts_handle_hid_data(struct ipts_info *ipts,
 
 	switch (raw_header->data_type) {
 	case TOUCH_RAW_DATA_TYPE_HID_REPORT: {
-		if (raw_header->raw_data_size_bytes > HID_MAX_BUFFER_SIZE)
-			raw_header->raw_data_size_bytes = HID_MAX_BUFFER_SIZE;
+		if (raw_header->raw_data_size_bytes > HID_MAX_BUFFER_SIZE) {
+			ipts_err(ipts, "input report too large (%lu bytes), skipping",
+				 raw_header->raw_data_size_bytes);
+			break;
+		}
 
 		memcpy(ipts->hid_input_report, raw_data,
 			raw_header->raw_data_size_bytes);
-- 
2.25.1

